metricq_source_http/source.py

```diff
@@ -0,0 +1,441 @@
+import asyncio
+import importlib
+import logging
+import logging.handlers
+import time
+from collections import defaultdict
+from contextlib import suppress
+from dataclasses import dataclass
+from functools import cache
+from typing import Any, Callable, Iterable, Optional, Sequence, cast
+
+import aiohttp
+import click
+import click_log  # type: ignore
+import metricq
+from hostlist import expand_hostlist  # type: ignore
+from metricq.logging import get_logger
+from yarl import URL
+
+from .version import __version__
+
+logger = get_logger()
+
+click_log.basic_config(logger)
+logger.addHandler(logging.handlers.SysLogHandler())
+logger.setLevel("INFO")
+logger.handlers[0].formatter = logging.Formatter(
+    fmt="%(asctime)s [%(levelname)-8s] [%(name)-20s] %(message)s"
+)
+
+
+class ConfigError(Exception):
+    pass
+
+
+@cache
+def load_plugin(name: str) -> Callable[..., float]:
+    try:
+        module = importlib.import_module(f"metricq_source_http.plugins.{name}")
+    except ImportError as e:
+        raise ConfigError(f"Could not load plugin {name}: {e}")
+    try:
+        return cast(Callable[..., float], module.response_parse)
+    except AttributeError:
+        raise ConfigError(f"Plugin {name} has no response_parse function")
+
+
+def _extract_interval(**config: Any) -> Optional[int | float]:
+    """
+    To allow interval, rate or maybe period, and other types in the future,
+    we have a separate function here.
+    """
+    with suppress(KeyError):
+        interval = config["interval"]
+        assert isinstance(interval, (int, float))
+        return interval
+    return None
+
+
+class AuthorizationManager:
+    _needs_login = False
+
+    @staticmethod
+    def create(
+        login_type: Optional[str] = None, **kwargs: Any
+    ) -> "AuthorizationManager":
+        if login_type is None or login_type == "none":
+            return AuthorizationManager()
+        if login_type == "cookie":
+            return CookieAuthorizationManager(**kwargs)
+        if login_type == "basic":
+            return BasicAuthorizationManager(**kwargs)
+        raise ConfigError(f"login_type {login_type} not supported")
+
+    @property
+    def session_params(self) -> dict[str, Any]:
+        return {}
+
+    async def authorize_session(self, session: aiohttp.ClientSession) -> bool:
+        """
+        Authorize the session and returns True on success.
+        If False is returned, you need not retry.
+        """
+        return False
+
+
+class BasicAuthorizationManager(AuthorizationManager):
```

Comment:
I guess this cannot work, as it's not overriding `authorize_session`, which means that `authorize_session` will always return `False`.

Reply:
I guess the `authorize_session` isn't quite as clear as I thought.
`authorize_session` is called in two cases:
1) At the beginning of the session, in which case the return value is ignored.
2) There was an error and maybe it's because of authorization. We try to authorize, and if that was successful, we try again. But since the basic auth cannot actually do anything, we return `False` to indicate that no retry is necessary.